lib/chef/resource/inspec_waiver_file.rb

```diff
@@ -0,0 +1,152 @@
+
+      property :expiration, String,
+        description: "The expiration date of the given waiver - provided in YYYY-MM-DD format",
+        callbacks: {
+          "Expiration date should match the following format: YYYY-MM-DD" => proc { |e|
+            re = Regexp.new('([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))').freeze
+            if re.match?(e) || e.nil?
+              true
+            else
+              false
+            end
+          },
+          }
+
+      property :run_test, [true, false],
+        description: "If present and true, the control will run and be reported, but failures in it won’t make the overall run fail. If absent or false, the control will not be run."
+
+      property :justification, String,
+        description: "Can be any text you want and might include a reason for the waiver as well as who signed off on the waiver."
+
+      property :backup, [false, Integers],
+        description: "The number of backups to be kept in /var/chef/backup (for UNIX- and Linux-based platforms) or C:/chef/backup (for the Microsoft Windows platform). Set to false to prevent backups from being kept.",
+        default: false
+
+      action :add do
+        filename = new_resource.file
+        yaml_contents = if ::File.file?(filename) && ::File.readable?(filename) && !::File.zero?(filename)
+                          IO.read(new_resource.file)
+                        else
+                          ""
+                        end
+        waiver_hash = {}
+        waiver_hash = ::YAML.safe_load(yaml_contents) unless yaml_contents.empty?
+        control_hash = {}
+        control_hash["expiration_date"] = new_resource.expiration.to_s unless new_resource.expiration.nil?
+        control_hash["run"] = new_resource.run_test unless new_resource.run_test.nil?
+        control_hash["justification"] = new_resource.justification.to_s
+
+        if waiver_hash.key?("#{new_resource.control}")
+          unless waiver_hash["#{new_resource.control}"] == control_hash
+            waiver_hash["#{new_resource.control}"] = {}
+            waiver_hash["#{new_resource.control}"] = control_hash
+            waiver_hash = waiver_hash.sort.to_h
+            file "Update Waiver File #{new_resource.file} to update waiver for control #{new_resource.control}" do
+              path new_resource.file
+              content waiver_hash.to_yaml
+              backup new_resource.backup
+              action :create
+            end
+          end
+        else
+          waiver_hash["#{new_resource.control}"] = {}
+          waiver_hash["#{new_resource.control}"] = control_hash
+          waiver_hash = waiver_hash.sort.to_h
+          file "Update Waiver File #{new_resource.file} to add waiver for control #{new_resource.control}" do
+            path new_resource.file
+            content waiver_hash.to_yaml
+            backup new_resource.backup
+            action :create
+          end
+        end
```

Comment:
It looks like the two branches of this conditional are the same except one uses `add` and one uses `update` in the description. Was that intentional? If so, there's probably a more succinct way to write this without repetition.